main.js

```diff
@@ -0,0 +1,250 @@
+        lastName: user.lastName
+    }
+}
+
+const newUser = getUserCopy(john);
+console.log(newUser.firstName);
+console.log(newUser.lastName);
+console.log(newUser === john);
+
+// 10
+
+function getAnimalDescription(animal){
+    return `This ${animal.color} ${animal.name} has ${animal.legs} legs.`
+}
+
+// 11
+
+const rooms = {
+    first: {
+        description: 'This is the first room',
+        items: {
+            chair: 'The old chair looks comfortable',
+            lamp: 'This lamp looks ancient'
+        }
+    },
+    second: {
+        description: 'This is the second room',
+        items: {
+            couch: 'This couch looks like it would hurt your back',
+            table: 'On the table there is an unopened bottle of water'
+        }
+    }
+}
+
+// 12
+
+function addContact(contacts, name, phoneNumber) {
+    const newContacts = {...contacts};
+
+    newContacts[name] = {
+        name: name,
+        phoneNumber: phoneNumber
+    }
+
+    return newContacts;
+}
+
+const contacts = {};
+const contactsWithJohn = addContact(contacts, 'John', '123 456 789');
+
+console.log(contactsWithJohn.John.name);
+console.log(contactsWithJohn.John.phoneNumber);
+console.log(contacts === contactsWithJohn);
+console.log(contacts.John);
+
+// 13
+
+function findContact(object, property) {
+    if (object.hasOwnProperty(property)) {
+        return `Name: ${property}, Phone: ${object[property].phoneNumber}`;
+    }
+    return;
+}
+
+const johnContactInfo = findContact(contactsWithJohn, 'John');
+console.log(johnContactInfo);
+
+// 14
+
+const userThree = {
+    firstName: 'Bob',
+    lastName: 'Ross',
+    age: 40,
+    greet() {
+        return "Hi, I'm Bob Ross!";
+    }
+}
+const objectExample = {
+    firstName: 'Kate',
+    lastName: 'Williams',
+    age: 40,
+    greet() {
+        return "Hi, I'm Kate!";
+    }
+}
+const car = {
+    make: 'Ferrari',
+    model: 'F40',
+    greet() {
+        return "<engine noise>";
+    }
+}
+
+function validateIfFirstNameisString(object) {
+    return object.hasOwnProperty('firstName')
+            && typeof object.firstName === 'string';
+}
+
+function validateIfLastNameisString(object) {
+    return object.hasOwnProperty('lastName')
+        && typeof object.lastName === 'string';
+}
+
+function validateIfAgeIsNumber(object) {
+    return object.hasOwnProperty('age')
+        && typeof object.age === 'number';
+}
+
+function validateIfGreetIsFunction(object) {
+    return (object.hasOwnProperty('greet')
+        && typeof object.greet === 'function');
+}
+
+function validateIfGreetContainsBothNames(object) {
+    return object.greet().includes(object.firstName)
+        && object.greet().includes(object.lastName);
+}
+
+function validateIfObjectIsAUser(object) {
+    return validateIfFirstNameisString(object)
+        && validateIfLastNameisString(object)
+        && validateIfAgeIsNumber(object)
+        && (validateIfGreetIsFunction(object) && validateIfGreetContainsBothNames(object));
```

Comment:
Personally, I think it would look a lot cleaner with additional variables.
```
const hasFirstName = typeof object.firstName === 'string';
const hasLastName = typeof object.lastName === 'string';
// ...
return hasFirstName && hasLastName ...
```